File: node/src/public-epod.ts
import { DEFAULT_CONFIG } from './http-client';
import type { ApiResult } from './index';

// ============ Types ============

export interface PublicSignDetail {
    tracking_no: string;
    recipient_name: string;
    delivery_address_summary: string;
    destination_country_code: string;
    policy_url: string;
    policy_version_hash: string;
    signature_level_required: string;
    allowed_proof_types: string[];
    signature_waived: boolean;
    expires_at: string;
}

export interface PublicConsentResponse {
    consent_id: string;
    policy_version_hash: string;
}

export interface PublicCaptureResponse {
    evidence_hash: string;
    status: string;
    hash_locked: boolean;
}

// ============ Client ============

/**
 * PublicEpodClient provides access to public EPOD signing endpoints.
 * These endpoints use token-based auth (sign_token) and do not require API keys.
 */
export class PublicEpodClient {
    private baseUrl: string;

    constructor(baseUrl: string = DEFAULT_CONFIG.baseUrl!) {
        this.baseUrl = baseUrl.replace(/\/$/, '');
    }

    async getSignDetail(signToken: string): Promise<ApiResult<PublicSignDetail>> {
        const response = await fetch(`${this.baseUrl}/api/v1/open/epod/sign/${signToken}`);
        if (!response.ok) throw new Error(`HTTP ${response.status}`);
        return response.json();
    }

    async getPolicy(signToken: string, lang = 'en'): Promise<ApiResult<any>> {
        const response = await fetch(`${this.baseUrl}/api/v1/open/epod/sign/${signToken}/policy?lang=${lang}`);
        if (!response.ok) throw new Error(`HTTP ${response.status}`);
        return response.json();
    }

    async recordConsent(signToken: string, consentTypes: string[], policyVersionHash: string): Promise<ApiResult<PublicConsentResponse>> {
        const response = await fetch(`${this.baseUrl}/api/v1/open/epod/sign/${signToken}/consent`, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({ consent_types: consentTypes, policy_version_hash: policyVersionHash }),
        });
        if (!response.ok) throw new Error(`HTTP ${response.status}`);
        return response.json();
    }

    async captureSignature(signToken: string, consentId: string, signatureData: string, proofType = 'signature'): Promise<ApiResult<PublicCaptureResponse>> {
        const response = await fetch(`${this.baseUrl}/api/v1/open/epod/sign/${signToken}/capture`, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({ consent_id: consentId, signature_data: signatureData, proof_type: proofType }),
        });
        if (!response.ok) throw new Error(`HTTP ${response.status}`);
        return response.json();
    }
}

File: rust/src/public_epod.rs
use crate::error::Result;
use crate::public_epod_types::{PublicCaptureResponse, PublicConsentResponse, PublicSignDetail};

pub struct PublicEpodClient {
    base_url: String,
    client: reqwest::Client,
}

impl PublicEpodClient {
    pub fn new(base_url: &str) -> Self {
        Self {
            base_url: base_url.trim_end_matches('/').to_string(),
            client: reqwest::Client::new(),
        }
    }

    async fn request_json(
        &self,
        path: &str,
        method: reqwest::Method,
        body: Option<serde_json::Value>,
    ) -> Result<serde_json::Value> {
        let url = format!("{}{}", self.base_url, path);
        let mut builder = self.client.request(method, &url);
        builder = builder.header("Content-Type", "application/json");
        if let Some(b) = body {
            builder = builder.json(&b);
        }
        let resp = builder.send().await?;
        let status = resp.status().as_u16();
        if !resp.status().is_success() {
            let msg = resp.text().await.unwrap_or_default();
            return Err(crate::error::ShipzyError::Http { status, message: msg });
        }
        let data: serde_json::Value = resp.json().await?;
        Ok(data)
    }

    pub async fn get_sign_detail(&self, sign_token: &str) -> Result<PublicSignDetail> {
        let path = format!("/api/v1/open/epod/sign/{}", sign_token);
        let data: PublicSignDetail = serde_json::from_value(self.request_json(&path, reqwest::Method::GET, None).await?)?;
        Ok(data)
    }

    pub async fn get_policy(
        &self,
        sign_token: &str,
        lang: &str,
    ) -> Result<serde_json::Value> {
        let path = format!(
            "/api/v1/open/epod/sign/{}/policy?lang={}",
            sign_token, lang
        );
        self.request_json(&path, reqwest::Method::GET, None).await
    }

    pub async fn record_consent(
        &self,
        sign_token: &str,
        consent_types: Vec<String>,
        policy_version_hash: &str,
    ) -> Result<PublicConsentResponse> {
        let path = format!("/api/v1/open/epod/sign/{}/consent", sign_token);
        let body = serde_json::json!({
            "consent_types": consent_types,
            "policy_version_hash": policy_version_hash,
        });
        let data: PublicConsentResponse = serde_json::from_value(self.request_json(&path, reqwest::Method::POST, Some(body)).await?)?;
        Ok(data)
    }

    pub async fn capture_signature(
        &self,
        sign_token: &str,
        consent_id: &str,
        signature_data: &str,
        proof_type: &str,
    ) -> Result<PublicCaptureResponse> {
        let path = format!("/api/v1/open/epod/sign/{}/capture", sign_token);
        let body = serde_json::json!({
            "consent_id": consent_id,
            "signature_data": signature_data,
            "proof_type": proof_type,
        });
        let data: PublicCaptureResponse = serde_json::from_value(self.request_json(&path, reqwest::Method::POST, Some(body)).await?)?;
        Ok(data)
    }
}
